astropy/utils/masked/core.py

```diff
@@ -1290,6 +1302,14 @@ def __format__(self, format_spec):
         else:
             return string
 
+    def __hash__(self):
+        # Try to be somewhat like a numpy array scalar if possible.
+        if not self.shape and not self.mask:
```

Comment:
In my opinion this is more readable. I also measured and it seems faster (marginally for scalars, but about 3x for large arrays)

```suggestion
        if not self.ndim > 0 and not self.mask:
```

Reply:
OK, changed to `self.ndim == 0`